hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/volume/MutableVolumeSet.java

```diff
@@ -429,52 +363,45 @@ private boolean addVolume(String volumeRoot, StorageType storageType) {
   }
 
   // Mark a volume as failed
-  public void failVolume(String dataDir) {
-    String hddsRoot = HddsVolumeUtil.getHddsRoot(dataDir);
-
+  public void failVolume(String volumeRoot) {
     this.writeLock();
     try {
-      if (volumeMap.containsKey(hddsRoot)) {
-        HddsVolume hddsVolume = volumeMap.get(hddsRoot);
-        hddsVolume.failVolume();
+      if (volumeMap.containsKey(volumeRoot)) {
+        StorageVolume volume = volumeMap.get(volumeRoot);
+        volume.failVolume();
 
-        volumeMap.remove(hddsRoot);
-        volumeStateMap.get(hddsVolume.getStorageType()).remove(hddsVolume);
-        failedVolumeMap.put(hddsRoot, hddsVolume);
+        volumeMap.remove(volumeRoot);
+        volumeStateMap.get(volume.getStorageType()).remove(volume);
+        failedVolumeMap.put(volumeRoot, volume);
 
-        LOG.info("Moving Volume : {} to failed Volumes", hddsRoot);
-      } else if (failedVolumeMap.containsKey(hddsRoot)) {
-        LOG.info("Volume : {} is not active", hddsRoot);
+        LOG.info("Moving Volume : {} to failed Volumes", volumeRoot);
+      } else if (failedVolumeMap.containsKey(volumeRoot)) {
+        LOG.info("Volume : {} is not active", volumeRoot);
       } else {
-        LOG.warn("Volume : {} does not exist in VolumeSet", hddsRoot);
+        LOG.warn("Volume : {} does not exist in VolumeSet", volumeRoot);
       }
     } finally {
       this.writeUnlock();
     }
   }
 
   // Remove a volume from the VolumeSet completely.
-  public void removeVolume(String dataDir) throws IOException {
-    String hddsRoot = HddsVolumeUtil.getHddsRoot(dataDir);
-
+  public void removeVolume(String volumeRoot) throws IOException {
     this.writeLock();
     try {
-      if (volumeMap.containsKey(hddsRoot)) {
-        HddsVolume hddsVolume = volumeMap.get(hddsRoot);
-        hddsVolume.shutdown();
+      if (volumeMap.containsKey(volumeRoot)) {
+        StorageVolume volume = volumeMap.get(volumeRoot);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `get()` may not be thread-safe at lines: **392 and 393**. In between the check and the `get()` another thread can remove the key and the `get()` will return `null`. The removes that can remove the key are at lines: **373**, **396**, and **425**.

**Fix**  
 Consider calling `get()`, checking instead of your current check if the returned object is `null`, and then using that object only, without calling `get()` again.

**More info**  
 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3Guard.java#L302-L304) (external link).